### ISSUE TITLE
Backpressure exposure for asynchronous send()

### ISSUE THREAD

**Issue by agoode** [WebAudio/web-midi-api#158]
We have so far been able to use synchronous `send` but this provides no mechanism for backpressure.

The current implementation in Chrome uses large internal buffers to hide this issue, but this is wasteful and results in the browser silently dropping messages when the buffer is overrun. This condition is not common with normal MIDI messages, but sysex messages can be of arbitrary length and can be slow to send, so Chrome also puts limits on those as well.

In order to allow implementations to avoid these buffers and arbitrary limits, we need a mechanism for non-blocking `send`.

**Comment by yutakahirano** [comment]
Is [Streams](https://streams.spec.whatwg.org/) useful for the use case? cc: @domenic @tyoshino

**Comment by domenic** [comment]
This does sound like a pretty good fit for a writable stream... In fact MIDIOutput looks pretty WritableStream-like in general. It is specifically designed to take care of this queuing and backpressure-signaling concern for you, so that other specs can build on top of it and reuse the architecture.

I'm not sure about MIDIInput, as I'm not familiar enough with MIDI. The choice would be between the current spec's no-backpressure, event-based model (where if you aren't subscribed you miss the event), vs. a readable stream model with the data buffered for future reading, and potential backpressure when it's not read.

Unfortunately the spec's current setup where MIDIInput and MIDIOutput both derive from MIDIPort doesn't seem very easy to fit with the stream API. I'm not sure how we'd do this without just creating a parallel WritableMIDIPort type (and maybe ReadableMIDIPort).

There's also the issue that [writable streams need some spec love](https://github.com/whatwg/streams/labels/writable%20streams), but if this is a urgent use case I can turn my attention to them and get that straightened out quickly.

**Comment by agoode** [comment]
The event-based model for MIDIInput is fine, since MIDI is essentially a multicast system, with no backpressure possible at the protocol level. MIDI doesn't guarantee reliable delivery.

Can WritableStream handle framing? We do parse the bytes and require that they are valid MIDI packets (1, 2, or 3 bytes in the common case, or arbitrary length in the case of sysex). Is this implementable as chunks in the streams standard? Can a stream reject an invalid chunk?

**Comment by domenic** [comment]
> Can WritableStream handle framing? We do parse the bytes and require that they are valid MIDI packets (1, 2, or 3 bytes in the common case, or arbitrary length in the case of sysex). Is this implementable as chunks in the streams standard? Can a stream reject an invalid chunk?

Yes, for sure. The underlying source's write hook (and optional writev hook, once we get that working) can process the data in arbitrary ways, returning a promise for when the processing is finished---which can be rejected to indicate that the input was invalid.

**Comment by jussi-kalliokoski** [comment]
:+1: for using streams from me.

We could add a new method to `MidiPort`: `Promise<Stream> openAsStream()` to preserve bw compat.

However, how do we solve sending timed messages?

**Comment by bome** [comment]
I'm somewhat reluctant for adding streams, it's seems like mainly adding clutter to the WebMIDI API and implementation, while providing very little new functionality. Can't we just add a function `sendAsync(data, timestamp)` with a Promise as return, or define a MIDIOutput listener which fires whenever a MIDI message is delivered?

> results in the browser silently dropping messages

That's really bad. In that case, the send() function should either block, or throw an exception of some sort (which would need to be defined in the spec).

**Comment by cwilso** [comment]
I'm REALLY uncomfortable rebasing on top of Streams, since it's highly
unlikely you would be piping any readablestream into a MIDI output device.
I think the best way to solve this is have send() return a Promise (that
resolves when it's been sent to the hardware), and expose the "current
available output message size" (i.e. the maximum chunk size you can write,
right now).  This would enable things that currently work to keep working;
 it would enable developers who want to send large amounts of sysex data to
do so, too.

**Comment by jussi-kalliokoski** [comment]
> I think the best way to solve this is have send() return a Promise (that
> resolves when it's been sent to the hardware), and expose the "current
> available output message size"

This would just be duplicating stream functionality, except in a more awkward and less interoperable way. I can see plenty of use cases for MIDI messages in Streams. For example:
- Developers can use general purpose stream transform utilities (such as buffering with different strategies like FIFO to ensure all messages get delivered in order, or another strategy that makes sure certain messages take priority).
- Pipe messages from one client's `MIDIInput` to another's `MIDIOutput` via P2P communications or websockets.

**Comment by domenic** [comment]
I agree that such a design is just duplicating streams and will need to reinvent the queuing mechanisms and so forth they are specifically designed to let other specs reuse. The "current available output message size" is further evidence of that kind of duplication (of the `desiredSize` property of a writable stream, after https://github.com/whatwg/streams/issues/318 is implemented).  It's similar to reinventing an IDBRequest-style object when promises are available, based on the speculation that nobody would ever need to reuse the return value inside a promise callback.

> However, how do we solve sending timed messages?

I'd assume each chunk would be of the form `{ timestamp, data }` or similar.

**Comment by jussi-kalliokoski** [comment]
> I'd assume each chunk would be of the form { timestamp, data } or similar.

That sounds reasonable.

**Comment by cwilso** [comment]
Can someone who understands streams put together a sketch of an API based on it, and examples of simple and complex usage?

**Comment by jussi-kalliokoski** [comment]
Maybe something like this:

``` webidl
interface MIDIOutput {
  // ...
  Promise<WritableStream> openAsStream({ size, highWaterMark });
}

interface MIDIInput {
  // ...
  Promise<ReadableStream> openAsStream({ size, highWaterMark });
}
```

here's [2 usage examples](https://gist.github.com/jussi-kalliokoski/9e6e1fd28c08b973cf2b) - the first is just a dumb non-synced sequencer writing to the first available `MIDIOutput`, the second one takes advantage of piping streams to pipe from first `MIDIInput` to first `MIDIOutput`, filtering out messages that aren't note on or note off, pushing all messages to channel 3, pitch shifting by 2 octaves and buffering based on number of messages at high water mark of `1000`.

EDIT: Note that the advanced example uses `TransformStream` which is not in the Streams spec yet.

**Comment by toyoshim** [comment]
Reply for the first Adam's description:
`send` is not synchronous at all. If so, we can just return boolean, but we cannot. Also generally said we should avoid having synchronous APIs as much as possible on the main thread. So, what we want to do in this thread is to add a reliable `send`-like method.

Maybe using Streams would be a right approach. But I feel it's a little complicated as Chris said. Also, I feel Web MIDI should be aligned with other similar modern APIs. For instance [Web Bluetooth](https://webbluetoothcg.github.io/web-bluetooth/) and [WebUSB](https://wicg.github.io/webusb/) return Promise.

So my preference is bome's `sendAsync` like thing. But, I feel `sendAsync` is confusing name since as I said existing `send` is also async.

Here is my proposal.
`send`: a user can use it for short messages, and sysex. But internal implementation may drop a message if a user want to send many data in a short time. Also very long sysex mesages would fail to send always.
`sendLong`: returning Promise. could not be called twice before the previous call finish. internal implementation never drop a message for buffer limits, but it does not ensure that the message reaches to the device because MIDI transport layer does not have any fail-safe mechanisms and the message could be dropped on immediate browser and OS shutdown. We probably need to allow sending long sysex messages as fragments.

**Comment by domenic** [comment]
The sketch in https://github.com/WebAudio/web-midi-api/issues/158#issuecomment-184102371 is pretty reasonable, although we'd update it to use the writer design. I'll add some explicit demonstrations of backpressure support:

``` js
destination.openAsStream().then(s => s.getWriter()).then(async (writer) => {
  console.log(writer.desiredSize); // how many bytes (or other units) we "should" write

  // note: 0 means "please stop sending stuff", not "the buffer is full and
  // we will start discarding data". So, desiredSize can go negative.

  writer.write({ data: data1, timestamp: ts1 }); // disregard backpressure

  // // wait for successful processing before writing more
  await writer.write({ data: data2, timestamp: ts2 });

  await writer.waitForDesiredSize(100); // wait until desiredSize goes >= 100

  writer.write({ data: oneHundredTwentyBytes, timestamp: ts3 });

  await writer.waitForDesiredSize(); // wait until desiredSize goes >= high water mark

  // also, maybe default timestamp to performance.now()?
});
```

I might also collapse @jussi-kalliokoski's `.openAsStream(new CountQueuingStrategy({ highWaterMark: 1000 }))` into `.openAsStream({ highWaterMark: 1000 })`.

**Comment by toyoshim** [comment]
After thinking about incremental writing of a large sysex, I noticed that it will allow a malicious attack to lock all output ports exclusively. E.g. just sending a "sysex start" byte will lock the port forever. So, we should keep on having the restriction that a user can not send an incomplete or a fragment message.

So even if we have a back-pressure, the sysex size will be limited to the maximum size of ArrayBuffer.

**Comment by toyoshim** [comment]
@domenic Have you ever talked with Web Bluetooth and WebUSB guys before?
It would be great if these all APIs include Web MIDI are consistent in terms of API design. So, if there was a discussion, I'd hear opinions discussed there.

**Comment by domenic** [comment]
In an offline thread @cwilso mentioned that @jussi-kalliokoski's examples are too complex and he'd like a three-liner. Here you go (slightly changed from the above since I am not sure why @jussi-kalliokoski made stream acquisition async):

``` js
const writer = midiOutput.asStream().getWriter();

await writer.write({ data: d1, timestamp: ts1 });
await writer.write({ data: d2, timestamp: ts2 });
```

**Comment by bome** [comment]
@toyoshim you can intersperse short messages during an ongoing sys ex message, and the implementation could also use a timeout to abandon stalled sys ex messages after a while.

Silently dropping MIDI messages is always bad, maybe the current `send()` function can return a bool: `false` on buffer full.

I agree that `sendAsync` is not good, but `sendLong` is also misleading. Also, why wouldn't it be possible to call your proposed `sendLong` function again before the previous one finished? I would welcome "unlimited" buffering there! So, maybe `sendBuffered`? or spelled out: `sendWithPromise`?

**Comment by domenic** [comment]
> Have you ever talked with Web Bluetooth and WebUSB guys before? It would be great if these all APIs include Web MIDI are consistent in terms of API design. So, if there was a discussion, I'd hear opinions discussed there.

I previously talked with Jeffrey about Web ... Bluetooth? ... and the conclusion was that since there was no backpressure support it wasn't urgent and we could wait on adding streams until later.

**Comment by agoode** [comment]
Stream acquisition is async since you could block for an arbitrary amount of time with an OS-level `open()` call.

**Comment by domenic** [comment]
> Stream acquisition is async since you could block for an arbitrary amount of time with an OS-level open() call.

That's fine; that just means that the first `write()` won't succeed (or fail) until the open comes back. The stream object can still be used, traditionally, in these scenarios.

**Comment by agoode** [comment]
Is there a way to force the `open` to complete, to ensure we don't have to wait until the `write` to determine success?

**Comment by domenic** [comment]
I don't understand what "force the open to complete" would mean. You can see an example implementation here if it helps: https://streams.spec.whatwg.org/#example-ws-backpressure (unfortunately the usage example still does not have writers and uses the coarse-grained `"waiting"` signal instead of the fine-grained `desiredSize` discussed above). The file descriptor gets opened immediately upon stream creation, but the stream machinery takes care (both in specs and in implementations!) of making sure not to write to the descriptor until that successfully completes.

**Comment by agoode** [comment]
So, we currently have the notion of `pending` in Web MIDI. This is the state where `open` has not yet fully completed. In streams, I think `pending` would mean `desiredSize = 0` and `open` would mean `desiredSize != 0`?

It's good to know when the MIDI port is fully open, since we can signal in a UI that "preflight" is complete and all devices are fully ready to send/receive.

**Comment by toyoshim** [comment]
@bome Web MIDI backend needs to multiplex all sending messages from multiple MIDIAccess and once one of it contains 'sysex start' byte, and it was sent to an actual device, we can not abandon stalled sysex in any way, right?

Making `send` return a boolean sounds possible, but it never means the data is sent to the device, but just means the data is successfully buffered.

The answer for the second question of only one request is we need to ensure message orders in cases where one of in-flight message fails in the middle. Please imagine a case there the second request fails asynchronously, and a user already send the third request that may success. That will cause something a user do not expect.

**Comment by agoode** [comment]
Let me introduce a classic MIDI use case, and folks can weigh in on various ideas for making it work.

An SMF file contains a stream of `<timestamp><event>` pairs (basically the same as your proposed chunk description above), which we want to stream to an underlying MIDI sink at exactly the correct time. MIDI itself has no concept of timestamps, so something has to schedule this. SMF files can trivially take arbitrary amounts of wall clock time to play, with arbitrary gaps between sounds.

Right now, Web MIDI clients have to schedule themselves to submit some number of events with timestamps, and use `setInterval` to remind themselves to send some more. We should be able to do better than this.

If we want to get fancy, allow for user-supplied tempo changes which take effect immediately, while the SMF is streaming.

**Comment by toyoshim** [comment]
@domenic hum... It probably makes sense that Web Bluetooth does not need Streams at this point. Bluetooth defines write ack in the protocol level, and OS level APIs seem to expose this model directly. So, mapping write and write ack to a write method with returning Promise sounds a straight-forward reasonable solution.

But, I believe WebUSB will need Streams more than Web MIDI does.

**Comment by toyoshim** [comment]
Here is my answer for async or sync `asStream`.

Probably we should make MIDIOutput explicitly require `MIDIOutput.open` completed before calling `asStream`. Since buffer allocation happens in renderer side, remaining tasks for `asStream` could finish synchronously.

**Comment by toyoshim** [comment]
For SMF playback, I'd prefer to use requestAnimationFrame(timestamp => { }) even though my task isn't related to graphics. We can calculate delta time in the callback, and send next messages that won't make it in the next callback cycle as estimated with the calculated delta.

**Comment by yutakahirano** [comment]
How about the close operation? I mean, what is the relation between `WritableStreamWriter.close` and  `MidiOutputPort.close`?

**Comment by toyoshim** [comment]
IMO, WritableStreamWrite.close would better to close the Streams pipe instance only, and does not close the actual MIDIOutput channel. On the other hand, if the MIDIOutput channel is closed, the Streams instance should be closed automatically and any in-flight operations will be aborted.

**Comment by cwilso** [comment]
Why does WebBluetooth (and WebUSB) lack backpressure?  Particularly considering those are two of the most common transports for MIDI, that seems untrue...

```
await writer.write({ data: d1, timestamp: ts1 });
```

What determines that object format?  And what does a corresponding input scenario look like?

> Right now, Web MIDI clients have to schedule themselves to submit some number of events with timestamps, and use setInterval to remind themselves to send some more. We should be able to do better than this.

No, we really, really should not.  If we do, we are creating a sequence system; you can certainly do that on TOP of what we're building here, but it should not be the low-level API.  In the same way that Web Audio defines a system you can build a sequencer on top of, but does not have a build in rescheduling capability, Web MIDI should not either.

As for building that sequencer - you shouldn't use requestAnimationFrame, because rAF is chained to when the page is visible - NOT what you want here.  In fact, the only reason setInterval is broken is because of the throttling of setTimeout/setInterval to 1Hz when the page is not in focus.  I raised fixing that (or more to the point, enabling a consistent setInterval somehow), but it's not there yet.  At any rate, rAF has particular semantics that are not desirable here.

**Comment by toyoshim** [comment]
@cwilso 
Web Bluetooth seems to expose low level packet transmission as is. BLE defines two kind of write operation for the Characteristic, one is with response and the other is without response. Usually maximum packet size is 20B or something. Allowed operations, maximum packet size, and so on are declared through the GATT service, characteristic by characteristic.

`BluetoothRemoteGATTCharacteristic` having `Promise<void> writeValue(BufferSource value);` is a straight-forward design that just exposes the underlying BLE API, and other well-known operating systems also expose similar asynchronous APIs. So, I feel it could be reasonable. I'm not sure if this is useful to understand this, but [here](https://github.com/toyoshim/KonashiOTAFU/blob/master/KonashiOTAFU/Downloader.swift) is one native application code to use BLE.

FYI, BLE MIDI spec requires `write without response` for MIDI OUT. So, my explanation can not answer your question though. MIDI IN uses `notify`. So, backpressure is not needed like Web MIDI.

USB also depends on underlying packet based protocol, but usually operating systems does not expose it as far as I know (but I didn't check very well). E.g., Linux USB driver is written by using traditional read/write like blocking APIs like [this](https://github.com/toyoshim/tree-status/blob/master/driver/linux/usb_crts.c). Exposing underlying blocking API to JavaScript as an asynchronous one is common challenge in the web standards. And I think WebUSB probably should have Streams for interrupt, bulk, and isochronous transfers because these are usually exposed as BiDi channels.

I'm strongly object to have a Streams for MIDIInput. That is useless and just makes browser implementation difficult. Also as Adam said, it does not fit to the broadcast messages.

But I'm neutral for MIDIOutput at this point. There would be pros and cons.

re: rAF
I assume that application stop playing on the background. Probably timer throttling could be a topic of [User Agent Intervention](https://github.com/WICG/interventions) that we already did. But it might be something that may require user's permission?

**Comment by agoode** [comment]
@cwilso Sorry I was a little unclear by "we" here. I agree Web MIDI should not provide a sequencer. But Web MIDI (or the Web in general) should provide some mechanism for implementing one. There are various partial solutions like setInterval or rAF, but they are not quite right as you point out.

What is the Web Audio mechanism? Do you mean AudioProcessingEvent? That is good for periodic data at a high rate, but MIDI data is aperiodic.

**Comment by toyoshim** [comment]
Why throttled 1 sec interval timer does not work for you?
You should be able to queue next sending messages for the next 1 sec with proper timestamps.
I don't think you need to adjust playback speed very quickly in a background tab, right?

**Comment by agoode** [comment]
1 second is probably ok, though I think you might get an audible glitch during the initial transition from 60Hz to 1Hz, when your buffer would empty early?

**Comment by notator** [comment]
> Why throttled 1 sec interval timer does not work for you?

My sequencer-like application sends timestamped midi data at irregular intervals of less than a second. When the tab goes into the background, all the messages that are timestamped within one second are played at the same time: At one second intervals. That's just ugly. In order to avoid this problem, the application has to stop itself playing when its tab goes into the background. It would obviously be better if the app could continue to play normally even when its tab is not at the front.

**Comment by toyoshim** [comment]
@agoode Ah, that's good point. I didn't investigate if there is a smart way to detect the timing to go background and change algorithm to play smoothly in JS without any glitch. But, if we didn't have, exactly [WICG](https://github.com/WICG/interventions) must be a right place to discuss to have a way. Now there have issues that discuss the timer throttling.

@notator You should be able to use fine-grained timestamp even if you are using 1sec interval timer. Originally the timestamp is introduced for such purpose, doing accurate playback in coarse-grained execution timing.

**Comment by notator** [comment]
> You should be able to use fine-grained timestamp even if you are using 1sec interval timer.

Can you point me at an example of how to do that? (EDIT: I probably misunderstood you there. Probably you meant that this is precisely what you are working on.)

BTW: like @bome, I'm also very interested in backpressure and overflowing buffers, so am following this thread with great interest!

**Comment by toyoshim** [comment]
Concept is something like this.

// Sending data in every 500ms within 1sec timer.
setInterval(() => {
 const now = performance.now();
 out.send(data, now + 0);
 out.send(data, now + 500);
}, 1000);

**Comment by notator** [comment]
Ach soooo... Thanks. But the details of doing that are not going to be easy. It would be much nicer if the problem could be solved in the browser once and for all, rather than having to be done by every app developer. Watching this space. Sorry for the interruption.

**Comment by cwilso** [comment]
I think James is using my polyfill, which uses setTimeout directly (i.e. I
always intended it to be a stopgap measure, and did not implement the async
scheduled send in a way that is resilient to the tab going into the
background).  The pattern of fine-grained timestamps is correct, but the
polyfill won't send them correctly in time if the tab goes into the
background.  Chrome's native implementation of Web MIDI will.

It's quite possible to do this, today - setTimeout and setInterval are not
throttled to 1Hz in Worker threads, only in the main thread.  So you can
use a worker thread to send yourself a timer message, and it will be
consistent.  (In fact, this is implemented in my metronome code at
https://github.com/cwilso/metronome.)  I should probably fix up the
polyfill.

The throttling was originally implemented by browsers to defeat naive use
of setTimeout for animation, which was causing horrendous power usage
(animation would happen even when offscreen, in short).  Although this was
done for good reasons, this particular case - where a developer really DOES
want a timer that occurs at >1Hz even while the tab is blurred - is a pain,
since you have to use a separate thread.  I'd proposed that we should have
a setTimer/setInterval switch that says "I know what I'm doing, don't
throttle me", but the use case is pretty narrow.

Incidentally, you can smoothly detect when the system switches between
full-resolution setTimeout and throttled setTimeout - just watch for the
window.onblur and window.onfocus events.  Of course, when the window comes
back into focus, you may have up to a second already scheduled, and you
can't get rid of it.

setInterval _is_ the right solution here, it just needs to be able to
disable the hack browsers put in to fix the naive animation power drain.

**Comment by notator** [comment]
@cwilso 

> I think James is using my polyfill, which uses setTimeout directly (i.e. I always intended it to be a stopgap measure, and did not implement the async scheduled send in a way that is resilient to the tab going into the background).  The pattern of fine-grained timestamps is correct, but the polyfill won't send them correctly in time if the tab goes into the background. Chrome's native implementation of Web MIDI will.

I'm using your polyfill for Firefox+Jazz, but Chrome's native implementation in Chrome. I just checked that this really is so by putting a breakpoint on the polyfill in both browsers.

> Chrome's native implementation of Web MIDI will.

You mean in future, right?

However, my app's "simple playback" code is still using a tick() function [1] that hasn't changed much since you practically wrote it for me in 2012. The function has a single call to setTimeOut, and does not use setInterval. It currently runs in the browser thread, but

> setTimeout and setInterval are not throttled to 1Hz in Worker threads, only in the main thread.

makes the solution look much closer. :-)

It seems that all I have to do is create a Worker thread that handles the performance timing and message sending.

My equivalent "assisted performance" code [2] marshalls midi messages that are prepared in Worker threads, but sends them in the main thread when the user hits the proper key.
Presumably I should also send those (marshalled) messages in a Worker thread, even though there's no user interaction when the tab is in the background. (The triggered events can be very long...)

Am I right in thinking that I needn't bother implementing these extra Workers, because the Web MIDI API is shortly going to provide alternatives that will keep my code simpler?

[1] in https://github.com/notator/assistant-performer/blob/master/ap/Sequence.js
[2] in https://github.com/notator/assistant-performer/blob/master/ap/Keyboard1.js

**Comment by cwilso** [comment]
@notator 

> > Chrome's native implementation of Web MIDI will.
> > You mean in future, right?

No, it should now.  There's no reason for the internal timestamping to be throttled.  I haven't personally tested this, but if it's not working it should get filed as a bug.

> It seems that all I have to do is create a Worker thread that handles the performance timing and message sending.

Actually, if you look at the metronome code I pointed at before, it has a nice module that does precisely that (sends a "tick" reliably to the main thread by using a worker thread setInterval).

**Comment by notator** [comment]
@cwilso I mentioned the problem a month ago at
https://bugs.chromium.org/p/chromium/issues/detail?id=497573#c9
but @toyoshim said

> that's expected Chrome behavior. Chrome throttles inactive tabs' timer invocations. It isn't Web MIDI specific something.

and

> Recommended solution is to stop playing when your tag is in background.

Which is why my app currently does just that. (Are we somehow talking at cross purposes?)

Liked your metronome code a lot!  I'll be implementing something along those lines if a simpler solution doesn't appear in the next few months...

This "background tab" issue seems unrelated to the backpressure/overflowing buffers problem, but maybe that's not the case...
My app can throw **lots** of data at the browser in a very short time, so I'm very interested to see how that situation can be controlled. In particular, I need to know when I start sending messages that are not going to get delivered.

**Comment by ryoyakawai** [comment]
@notator 
Why do you need that browser's background tabs run in same priority as the foreground tab?
I am just curious because I usually design and implement applications which run everything within one tab (some times in several windows.)

I guess the one of the reasons why Chromium/Chrome force background tabs to be lower priority is because to make less consumptions of the hardware resources(Battery, CPU, Memory, etc) as much as possible. And I think to make decision of browser's priorities is the browser's policy, so it is not good idea that those priority is controllable by each API's issues, I think.

**Comment by cwilso** [comment]
All modern browsers throttle setInterval/setTimeout to 1Hz because a lot of developers used them to do animation at 60Hz, and didn't bother to shut them down when the tab when out of focus.  It caused amazing background-tab battery drain.

**Comment by notator** [comment]
@ryoyakawai This actually has a fairly low priority for my app, but I think it would be more web-consistent if browsers didn't throttle Web MIDI apps when they are running in a background tab.

Apps that produce sound, but don't use setInterval/setTimeout -- such as Flash players, YouTube videos, html `<audio>` tag (? I haven't tested that) etc. don't get throttled or stopped when they run in the background. And Web MIDI apps get throttled when running in a background _tab_ but not in a background _window_. That's going to confuse users unless the app programmer goes to the extra trouble of remedying the problem with a solution like Chris' metronome. It would be _much_ better if Web MIDI apps behaved web-consistently by default.

Maybe we simply need a Web MIDI replacement for setTimeout that doesn't get throttled? I'm not currently using setInterval, and don't think I'd ever need to if we had an unthrottled replacement for setTimeout.

By the way: I said

> My app can throw **lots** of data at the browser in a very short time, so I'm very interested to see how that situation can be controlled. In particular, I need to know when I start sending messages that are not going to get delivered.

More precisely: I need to know **before** I send a message whether its going to be delivered or not.

EDIT: It looks as if we have two different issues here: unthrottled setTimeout _and_ the "mechanism for backpressure" that involves `send`. Maybe these issues are related after all?

**Comment by toyoshim** [comment]
Browsers introduced such throttling to protect users from selfish applications. Also, timer throttling is not breaking  compatibility, but allowed by spec AFAIK.
Adding unlimited version of setTimeout/Interval would be considerable, but it might be something allowed under an user's permission. Anyway, I think it's out of scope in this thread and WG, and should be discussed in the right place, WICG. ([Issue list](https://github.com/WICG/interventions/issues))

Web MIDI's timestamps should not be throttled as Chris said, and it works or worked on all platforms. But there were big changes on thread and task schedulers inside Chrome. So, it's possible that there is a platform dependent regression. If it does not work as intended on your platform, please file a bug with a simple sample code that reproduce the issue from crbug.com.

**Comment by notator** [comment]
In spite of what toyoshim just said, I'd like to record a thought I had this morning. Maybe there's a mistake in there somewhere, but anyway:

If we had a new, (non-throttling) `setTimeout` in the Web MIDI API (`setMidiTimeout`), then it could be designed to solve the backpressure issue as well. It could return the size of the currently available buffer. If I'm worried that my messages are going to overflow the buffer, I could then do something like this:

``` javascript
var remainingBufferSpace = setMidiTimeout(undefined, 0, 0);
var message = nextMessage(); // the next message that tick() is going to send
var nextMessageLength = length(message);
while(thereAreStillMessagesToBeSent)
{
    if(remainingBufferSpace > nextMessageLength)
    {
        // send the message (at least load it into the buffer successfully)
        remainingBufferSpace = setMidiTimeout(tick, delay, nextMessageLength);
        message = nextMessage();
        nextMessageLength = length(message);
    }
    else
    {
        ...
    }
}
```

And there would be no need to change the way `send` behaves.
EDIT1: fixed a bug in the above code.
EDIT2: added a third argument to `setMidiTimeout`, so that it can subtract the value from its current buffer size and return immediately.

**Comment by cwilso** [comment]
There are too many missing bits of information for me to fully grok your example (e.g. the undefined parameter, the tick, how a while loop in the main thread doesn't turn into a synchronous mess).  However, I feel VERY strongly that we do not need a new MIDI timeout API that calls back into the main UI thread - that's setTimeout.  Yes, it has an unfortunate throttling - however, 1) that problem should be solved universally for setTimeout/setInterval users that know what they're doing, via a flag or some such, and 2) there's a fairly trivial workaround, which is to use worker threads (seriously, read through the metronome code; I pretty much packaged the worker tick as a drop-in to any other sequencer).  You should not have the problem you describe, UNLESS there is a bug in Chrome's timestamped-future-send of MIDI messages with the new scheduler, if you have properly used the worker workaround.

(Note that it's relatively easy to switch back and forth on window.blur and window.focus, too, so you don't even always need a worker thread running.

**Comment by notator** [comment]
Understood. I was just brainstorming, and in a bit of a hurry...

Is the setTimeout issue already being dealt with in the proper places?

Your metronome works fine for me in a background tab in Chrome, so I assume there is no bug in "Chrome's timestamped-future-send of MIDI messages".

As far as I'm concerned, the backpressure issue could be solved if `send` would simply throw an exception of some sort when the message is too big for the available buffer space. Is there a standard `BufferOverflow` exception ? ( I have been unable to discover where the Web MIDI API exception types are defined.)

**Comment by toyoshim** [comment]
As I repeatedly said, WICG is the right place to discuss such throttling, and there are threads for the setTimeout throttling in some cases, e.g., background tab, and cross origin iframe. Actually there is more aggressive proposal to suspend background tabs completely. IMO, current Web MIDI use case isn't fair enough to avoid such throttling.

Exception does not solve the problem because users can not be notified when buffer is available again. Retry with setTimeout won't be an acceptable choice.

**Comment by notator** [comment]
> Exception does not solve the problem because users can not be notified when buffer is available again. Retry with setTimeout won't be an acceptable choice.

But I'd like to be told when the buffer overflows, even if I can't recover.

**Comment by cwilso** [comment]
To be clear - the setTimeout issue is not relevant here.  There's a workaround - the worker setTimeout, as I use in the metronome (https://github.com/cwilso/metronome/commit/7b78d49d297107bec2c10803ee2702f5b4e9e3cc).  No one seemed particularly interested in fixing the higher-level issue, but that would land in DOM-space somewhere.

The issue here is 1) how we expose backpressure for sending data to MIDIOutput, in order to enable "large" sends, and 2) whether we use Streams to do that (and a corollary, whether there's any significant utility in adding Streams to MIDIInput then.

I feel VERY strongly, btw, about not removing send() as-is.  I also think SOME backpressure fix should go in v1.

**Comment by notator** [comment]
+1 for Chris' summary.
Maybe a couple more scenarios would help (its not just a SysEx problem):

Scenario 1:
a) An app allows its users to change the speed at which a MIDI file is played back.
b) A user plays the file so fast that data starts getting lost.
c) The user doesn't know this, and goes faster and faster. Info is being silently ignored. This means that patch-changes, noteOns, noteOffs etc. are getting lost randomly, and the file is going to sound wrong. There's no point in continuing in this situation, so the app ought to be able to stop it.
The app ought to be able to tell the user that the file can't be played that fast. _Ideally_, the user should be told that _before_ the app starts sending the too fast messages.

Scenario 2:
A composer is creating a MIDI composition containing large amounts of data being sent very fast. The problem is that there is no way of knowing that the buffer is overflowing, except that each performance is going to sound randomly different from the one before. Having a professional attitude, the composer wants to use the browser's capabilities to the full, but still be sure of having predictable performances. There's no point in writing piccolo music for a tuba.
This scenario also raises the question of garbage collection. Can compositions of arbitrary length be performed without pauses for GC?

**Comment by cwilso** [comment]
WG decision: move backpressure to v2 incubation; streams is the likely right API.

**Comment by starfishmod** [comment]
Hi I wanted to re-raise the awareness of this issue. I am currently developing prototypes for some of the new MIDI-CI specifications. A large part of what is required is to receive and send lots of SysEx packets at one time. 
As I am using USB MIDI my bandwidth can be much higher than 31.25 kbps.   
Currently I can receive SysEx at high rates at the speed of whatever the USB MIDI sends through it's data.
When it comes to sending data I have to hobble the speed to 31.25Kbps (using the send with timestamp) as I can't determine the bandwidth.
And if I send the data too quickly then packets are silently dropped and the receiving device misses vital packets.
I know this is slated for v2 but can this be reviewed sooner. I am happy to discuss any details I can.
- Andrew

**Comment by toyoshim** [comment]
I'm revisiting here during Web MIDI bug triage for Chrome.

So, I just want to say that chromium side bug for this topic is here, https://crbug.com/383578, and we will work on that once something is decided here, in the spec side thread.

Note, here is the right place to discuss the API surface. Please do not discuss it at the crbug, just in case.

**Comment by starfishmod** [comment]
@toyoshim what can we do to help?
Is it as simple as using send and receiving a buffer is full error? and therefore go to a queue?
Should SysEx use a different buffer?
What are the (good) possible options that should be considered?
thanks for looking into this :)

**Comment by toyoshim** [comment]
@cwilso 's last comment is a current plan.
I agreed that using Streams would be a reasonable idea to solve this kind of issues.

When we discussed last time, Stream was not stable yet maybe?
Also, any other UAs didn't have much progress on implementation. So, it was marked as V2.

But both situations have been changed. We may be able to revisit this for V1. I'm fine if we go.

@padenot or @qdot may have some thoughts from Mozilla

**Comment by solitud** [comment]
I am not competent to discuss but I am wondering if there is a solution or workaround to send larger SysEx dumps with Web Midi? I am maintaining a SysEx librarian under https://f0f7.net and I have problems to send bulk dumps (~11kb) to older synths like MicroWave1. It seems like Web Midi is sending data faster than some interfaces or synth can process which results in buffer overflows/data loss.
I tried to send the data in timestamped chunks but than of course I get *System exclusive message is not ended by end of system exclusive message*
I am guessing regarding to the age of this thread it will need some more time to have full working SysEx buffer size/speed support in Web Midi so I would really appreciate if someone can point out if there is a current workaround to slow down the transmission?

**Comment by toyoshim** [comment]
I don't think 11kB is not so large to be handled in Chrome Web MIDI implementation.
On Windows, Chrome Web MIDI backend has 256kB limit per each SysEx message size, and also all platforms including Windows has 10MB limit for in-flight buffer size.

As far as I know, only sampling data transfer would hit this limit, and I heard it can be split into multiple SysEx messages.

At this moment, what users can do is to estimate the worst case throughput, then send split messages step by step instead of sending it at once. E.g., using setInterval(), and send each 128kB in the timer callback would work. Of course, each message should be a complete SysEx message as you can see the error.

**Comment by agoode** [comment]
Note that for Linux specifically, there is a long standing kernel bug that silently drops parts of most sysex messages unless the writer explicitly writes slowly. Unfortunately, Chrome uses the broken API and this will cause problems on Linux and Chrome OS devices. I suspect other Linux Web MIDI implementations will also use the ALSA seq API which has this problem.

I would like to fix this Linux bug but have not yet. Some more background can be found at http://crbug.com/917708.

**Comment by inno** [comment]
How can we get some traction on finalizing this?

Limiting buffer length, either via bug or definition, is not aligned with the MIDI specification. There is actually _no_ formal expectation of use or length of sysex messages. Per MIDI 1.0 96.1v3: 

> System Exclusive messages ... are used to transfer **any number** of data bytes in a format specified by the referenced manufacturer.

If folks are looking for a real world example of an arbitrarily large sysex message, I have two words: _firmware updates_. Different manufacturers handle these different ways, but many simply have "large" encoded binaries and checksums in a single message. Manufacturers are allowed to do this, but unfortunately, not required to handle chunking this data in any way (by spec).

Happy to provide more info and/or obscure use cases if needed!

**Comment by toyoshim** [comment]
When we started this discussion, Streams API was not matured yet.
But today, having WritableStream in MIDIOutput interface sounds the best choice?
We already have many tips to write Promise-based asynchronous code in a simple way, that says adding another asynchronous trick won't be a good idea.

We prefer to keep the existing send() method together for compatibility and simple uses. That sounds reasonable to me.

@cwilso Any thoughts? You were sceptical about using Streams in the past, but WDYT today?

I'm happy to implement it in Chrome if the spec is updated.

**Comment by diegodorado** [comment]
What is the current status of this issue?
Any custom code example to implement meanwhile?

**Comment by toyoshim** [comment]
Discussed at TPAC, and agreed that using Streams is the best here, but still we need to clarify detailed API surface.

**Comment by ghost** [comment]
I built a midiWebApi sequenser that work to playback,,playup, record "simultaneously" well more or less. But the playup lags because i use timeout, i would like to change the timeout solution to a buffered playup. Will it remove the playup lag?
Can you direct me to an example on howto implement a buffered playup, i do not quite get what it is. Since i just use timeouts/waitstate beteween note playup.

Very few examples showing how buffered playup work.

https://jonasth.github.io/

**Comment by cwilso** [comment]
Hey @jonasth, I wrote an article a long time ago about Web Audio clocks, that also applies to Web MIDI - in particular, why using setTimeout is not a good idea.  https://www.html5rocks.com/en/tutorials/audio/scheduling/.

The issue you're having is unrelated to this issue (this isssue is essentially about sending massive amounts of data on a slow MIDI link).

**Comment by ghost** [comment]
But i heard webmidi players that i at least "perceive" as in synch not slowing down passing notes, sure  enough there is no bar to measure if they lag or not so i really don't know, but you claim it is webMidi itself create the bottleneck for passing notes?

Thanks for your kind reply, seldom i get any.

Jonas
________________________________
Från: Chris Wilson <notifications@github.com>
Skickat: den 26 februari 2021 20:41
Till: WebAudio/web-midi-api <web-midi-api@noreply.github.com>
Kopia: jonasth <jonas.thornvall@hotmail.com>; Mention <mention@noreply.github.com>
Ämne: Re: [WebAudio/web-midi-api] Backpressure exposure for asynchronous send() (#158)


Hey @jonasth<https://github.com/jonasth>, I wrote an article a long time ago about Web Audio clocks, that also applies to Web MIDI - in particular, why using setTimeout is not a good idea. https://www.html5rocks.com/en/tutorials/audio/scheduling/.

The issue you're having is unrelated to this issue (this isssue is essentially about sending massive amounts of data on a slow MIDI link).

**Comment by ghost** [comment]
function STARTPLAY(){
console.log("STARTPLAY()");
copyEv = playEv.slice();
last = copyEv[copyEv.length - 1]
copyEv[copyEv.length] = last;
var waittime = copyEv.shift();
console.log("waittime="+waittime);
midEvent = 0;  myTime=waittime;
//console.log(SF2PLAY);
if (SF2PLAY==true){stopJump=setTimeout(SF2Playup, waittime);}
else {stopJump=setTimeout(doPlayup, waittime);}
}
//WHILE MESSAGES TO PLAY DO PLAYUP
function doPlayup() {
if (keepGoing) {
if (copyEv.length) {
if (track[playtrack].midiMess[midEvent].data0<192 || track[playtrack].midiMess[midEvent].data0>207)
{noteMessage = [track[playtrack].midiMess[midEvent].data0, track[playtrack].midiMess[midEvent].data1, track[playtrack].midiMess[midEvent].data2];
} else { noteMessage = [track[playtrack].midiMess[midEvent].data0, track[playtrack].midiMess[midEvent].data1];}
if (mode=="Play"){
pianoKeypressOut(); scrollPianoOut();
}
outportarr[outportindex].send(noteMessage);
waittime = copyEv.shift();
midEvent++;
setTimeout(doPlayup, waittime);
} else { stopPLAY();}
}
}

This is the playup function



________________________________
Från: Chris Wilson <notifications@github.com>
Skickat: den 26 februari 2021 20:41
Till: WebAudio/web-midi-api <web-midi-api@noreply.github.com>
Kopia: jonasth <jonas.thornvall@hotmail.com>; Mention <mention@noreply.github.com>
Ämne: Re: [WebAudio/web-midi-api] Backpressure exposure for asynchronous send() (#158)


Hey @jonasth<https://github.com/jonasth>, I wrote an article a long time ago about Web Audio clocks, that also applies to Web MIDI - in particular, why using setTimeout is not a good idea. https://www.html5rocks.com/en/tutorials/audio/scheduling/.

The issue you're having is unrelated to this issue (this isssue is essentially about sending massive amounts of data on a slow MIDI link).

**Comment by rianhunter** [comment]
What is the status of this issue? Without any way to detect backpressure I have to artificially slow down sending bursts of messages to account for older midi devices that don't use USB.

For what it's worth I agree with @bome:

> I'm somewhat reluctant for adding streams, it's seems like mainly adding clutter to the WebMIDI API and implementation, while providing very little new functionality. Can't we just add a function `sendAsync(data, timestamp)` with a Promise as return, or define a MIDIOutput listener which fires whenever a MIDI message is delivered?

A method like his proposed `sendAsync` would be sufficient, extensible, and relatively easy to spec.

**Comment by gudsoft** [comment]
I second the other requests for providing a solution for this issue.  It is important.  For many of the reasons already discussed in this thread.

First, I would like to commend all the work done on Web MIDI API so far...  it is truly exciting (to me) what browser-based MIDI means!  At the same time, and at this stage of the development of the API, it is quite surprising this foreseeable and basic reality (the need for SysEx throttling) has not yet been accommodated by the API (and is the subject of so much debate and "needed explanation").  The need for throttling is simply a reality of the mature MIDI ecosystem and its legacy, much of which is just as relevant today as it ever was.  Pointing fingers at old or new devices, drivers, etc. (or wherever the need for throttling may arise in a particular system) does very little to advance all the good that the Web MIDI API otherwise brings.

Every piece of MIDI software of consequence created in the last few decades that deals with SysEx has recognized that providing an ability to throttle SysEx transmission is a necessity, either to accommodate limitations of the receiving device or transmitting system (OS, drivers, interfaces, etc. -- and now potentially even arbitrary internal browser limitations!).  Indeed, the computer, with its infinite flexibility for adaptation and reprogramming, is in the *best possible position* to accommodate "harder" limitations found elsewhere in a MIDI system (within "set-in-stone" old hardware, or "hard-to-update" firmware, etc.).  As a component of that flexibility, the Web MIDI API very clearly has a role to play here, IMO.

The Web MIDI API should provide (or at least enable) a solution for throttling SysEx transmission, period.  If it does not, whole classes of "imperfect", yet still valuable and viable, MIDI devices and software (synthesizers, interfaces, and even apps, etc.) will be needlessly cut out of the benefits of browser-based MIDI.

An arguably important "modern" case in point:  Many of the cheap "commodity" USB MIDI cables available today (which usually have full-speed SysEx transmission bugs) could benefit (read, be made to "actually work") via a simple adaptation of the software running on the computer (to throttle SysEx).  Wouldn't it actually further the goals of the Web MIDI API to *enable* as many MIDI devices (and users!) as possible, such as these commodity interfaces, despite their flaws?

Web MIDI API's current lack of any means for SysEx throttling, as well as its insistence that whole SysEx messages be provided for transmission are, IMO, arbitrary and harmful limitations of the API, and prohibit certain valid and creative uses of MIDI.  For example, why can't my browser-based MIDI app "stream" SysEx (generating it over time), or "open" a bidirectional SysEx connection using a single 0xF0 on each side, and then communicate asynchronously entirely with an "exclusive" protocol of 7-bit messages?  Both things are trivially possible with MIDI itself, but currently not the Web MIDI API.  (It should be pointed out, some of the earliest -- and perfectly valid -- SysEx implementations in Casio and Roland devices are effectively asynchronous 7-bit protocols, reliant on intra-SysEx handshaking, something the Web MIDI API is currently incompatible with, for no good reason, IMO.)

MIDI has grown to its popularity and ubiquity today, yes, based on the efforts of many to adhere to a robust, simple, and dare we say "perfect" common specification, but also in no small part due to the creative (and arguably "necessarily obvious") accommodations that have been made, especially in software, aimed at making sure MIDI works as well as possible even for some of the most "imperfect" members of the ecosystem, and that unforeseen innovations are not precluded by assumed "valid uses" of features like System Exclusive.

It would be awesome if the Web MIDI API wholeheartedly recognized and continued in this tradition.

**Comment by gudsoft** [comment]
As a separate note, I would like the suggest that, the `sendAsync` method proposed earlier in this thread should not only allow an individual SysEx message to be sent asynchronously, with completion signaling via `Promise` or other mechanism, but it should allow a *partial* SysEx message to be sent (nominally to be followed by the sending of subsequent part(s) until completion of the whole message).

Adding these two basic abilities...  *asynchrony* and *divisibility*...  are enough to "re-enable" everything that is natively possible with SysEx over MIDI itself, and would provide the means for the Web MIDI API to fully accommodate past and future innovations, and other vital use-cases that have been with us since MIDI's birth (see below).  A suitable `sendAsync` would accomplish this, as would some kind of streaming interface, but the important thing is that the API expose both abilities *together*.

It should also be pointed out that SysEx throttling isn't the only issue of importance here...  the ability for apps to freely interject MIDI realtime messages (such as `Timing Clock`) during a long-running SysEx message (something the MIDI protocol has always been *designed* to handle) is an important use-case for playback and synchronization.  Again, fine control over the *divisibility* of SysEx data from the perspective of the API is required.

I applaud any and all principled efforts to keep the Web MIDI API simple and easy-to-use (achieving this for any API is no small feat), but for all the reasons above, a synchronous, atomic `send` method is, in fact, *too simple*, and precludes much that the API could and should enable for browser-based MIDI apps.

**Comment by hoch** [comment]
2023 TPAC Audio WG Discussion:
The WG is still undecided about whether this feature will be included in V1 or the next version of the specification. It is likely to be added later, after developer surveys, prototyping, and field experiments have been conducted.

**Comment by mjwilson-google** [comment]
I will schedule this for CR / V1 until we have a chance to discuss it in the Audio Working Group, but we will likely push it to future work.

**Comment by mjwilson-google** [comment]
Audio Working Group 2023-10-05 meeting conclusions:
- This looks like it will be a significant amount of spec and implementation work
- It also looks like it is important
- The Working Group doesn't have enough information to make a final decision on how to handle this, so we will conduct a developer survey

**Comment by rianhunter** [comment]
Thanks @mjwilson-google looking forward to seeing progress on this issue!

**Comment by mjwilson-google** [comment]
TPAC 2024 notes:

This is actually somewhat covered by MIDI 2.0, which is out of scope for version 1 of the Web MIDI specification.

I propose moving this to future work, and not fixing this version 1.0 of the specification.

**Comment by agoode** [comment]
Does MIDI 2.0 contain some kind of flow control within the protocol? If so, that is different than this proposal, which is meant to reflect the backpressure that the existing MIDI 1.0 APIs already provide (by blocking at the OS level).

**Comment by starfishmod** [comment]
MIDI 2.0 UMP does not contain a low-level flow control. There are some flow control mechanisms within MIDI-CI.
Maybe @mjwilson-google is referring to the fact that UMP messages (inc SysEx7) is in packets now. So an application is not sending large buffers of SysEx all at once.
However, an application may be sending a large number of UMP's at one time and a similar buffering issue may occur?

**Comment by cwilso** [comment]
This shouldn't be related to MIDI 2.0; this was a problem reported by real-world use cases.  (When sending sysex, typically.).

**Comment by mjwilson-google** [comment]
I did not check details in the MIDI 2.0 spec.  I am seeing strong objections to pushing this out so we'll keep it scheduled for the current CR milestone.  But I still don't have a good idea of how to specify this yet, or how to conduct a valid survey.

I think I will try to fix the other CR issues first.  If this becomes the last blocker it will be easier to focus on.  I am aware it may take some time to resolve this issue which is motivation for trying to get things moving sooner, but it also seems like it's possible to make changes during CR review and it should be easier to get more eyes on the spec during that process.  We may end up drafting a change that we have low confidence in, and use the wide review to help verify if it's sufficient or not.

Thanks for the quick responses, and more feedback is always welcome.

**Comment by cwilso** [comment]
Happy to brainstorm a solution with you, @mjwilson-google

**Comment by mjwilson-google** [comment]
TPAC 2025 discussion:
- We can basically copy the mechanism in WebCodecs: see https://w3c.github.io/webcodecs/#saturated
